// @webpack-cli/utils/lib/run-prettier.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runPrettier = void 0;
const fs_1 = __importDefault(require("fs"));
const prettier_1 = __importDefault(require("prettier"));
const logger_1 = __importDefault(require("webpack-cli/lib/utils/logger"));
/**
 *
 * Runs prettier and later prints the output configuration
 *
 * @param {String} outputPath - Path to write the config to
 * @param {Node} source - AST to write at the given path
 * @returns {Void} Writes a file at given location
 */
function runPrettier(outputPath, source) {
    let prettySource = source;
    try {
        prettySource = prettier_1.default.format(source, {
            filepath: outputPath,
            parser: 'babel',
            singleQuote: true,
            tabWidth: 1,
            useTabs: true,
        });
    }
    catch (err) {
        logger_1.default.warn(`\nWARNING: Could not apply prettier to ${outputPath} due to validation error, but the file has been created`);
        prettySource = source;
    }
    return fs_1.default.writeFileSync(outputPath, prettySource, 'utf8');
}
exports.runPrettier = runPrettier;
